storage: TestRangeTransferLease failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/26a06a661e7c1ce63f3077269005472182286ca4

Parameters:
```
TAGS=deadlock
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=596408&tab=buildLog

```
W180411 09:11:43.859027 2103043 storage/raft_transport.go:285  unable to accept Raft message from (n2,s2):2: no handler registered for (n1,s1):1
```

```
[14:04:34] :	 [Step 2/2] --- FAIL: TestRangeTransferLease (10.23s)
[14:04:34] :	 [Step 2/2]     --- PASS: TestRangeTransferLease/Transfer (0.01s)
[14:04:34] :	 [Step 2/2]     --- PASS: TestRangeTransferLease/TransferWithExtension (0.01s)
[14:04:34] :	 [Step 2/2]     --- PASS: TestRangeTransferLease/DrainTransfer (5.01s)
[14:04:34] :	 [Step 2/2]     --- PASS: TestRangeTransferLease/DrainTransferWithExtension (0.01s)
[14:04:34] :	 [Step 2/2] 	leaktest.go:110: Leaked goroutine: goroutine 2082706 [sleep]:
[14:04:34] :	 [Step 2/2] 		time.Sleep(0x12a05f200)
[14:04:34] :	 [Step 2/2] 			/usr/local/go/src/runtime/time.go:102 +0x166
[14:04:34] :	 [Step 2/2] 		github.com/cockroachdb/cockroach/pkg/storage.(*Store).SetDraining(0xc42ef35c00, 0x1)
[14:04:34] :	 [Step 2/2] 			/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1074 +0x255
[14:04:34] :	 [Step 2/2] 		github.com/cockroachdb/cockroach/pkg/storage_test.TestRangeTransferLease.func11.2(0xc442dbdba0)
[14:04:34] :	 [Step 2/2] 			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_replica_test.go:826 +0x48
[14:04:34] :	 [Step 2/2] 		created by github.com/cockroachdb/cockroach/pkg/storage_test.TestRangeTransferLease.func11
[14:04:34] :	 [Step 2/2] 			/go/src/github.com/cockroachdb/cockroach/pkg/storage/client_replica_test.go:825 +0x242
```

Test flake, introduced in #22767 which introduces a five second sleep in that code path on `(Store).SetDraining` to wait for leases to transfer away. Putting a random 5s sleep in the quit pah is awkward. We should do better.

--- a/pkg/storage/client_replica_test.go
+++ b/pkg/storage/client_replica_test.go
@@ -490,147 +490,222 @@ func TestRangeLookupUseReverse(t *testing.T) {
 	})
 }
 
-func TestRangeTransferLease(t *testing.T) {
-	defer leaktest.AfterTest(t)()
+type leaseTransferTest struct {
+	mtc                        *multiTestContext
+	replica0, replica1         *storage.Replica
+	replica0Desc, replica1Desc roachpb.ReplicaDescriptor
+	leftKey                    roachpb.Key
+	filterMu                   syncutil.Mutex
+	filter                     func(filterArgs storagebase.FilterArgs) *roachpb.Error
+	waitForTransferBlocked     atomic.Value
+	transferBlocked            chan struct{}
+}
+
+func setupLeaseTransferTest(t *testing.T) *leaseTransferTest {
+	l := &leaseTransferTest{
+		leftKey: roachpb.Key("a"),
+	}
+
 	cfg := storage.TestStoreConfig(nil)
 	// Ensure the node liveness duration isn't too short. By default it is 900ms
 	// for TestStoreConfig().
 	cfg.RangeLeaseRaftElectionTimeoutMultiplier =
 		float64((9 * time.Second) / cfg.RaftElectionTimeout())
-	var filterMu syncutil.Mutex
-	var filter func(filterArgs storagebase.FilterArgs) *roachpb.Error
 	cfg.TestingKnobs.EvalKnobs.TestingEvalFilter =
 		func(filterArgs storagebase.FilterArgs) *roachpb.Error {
-			filterMu.Lock()
-			filterCopy := filter
-			filterMu.Unlock()
+			l.filterMu.Lock()
+			filterCopy := l.filter
+			l.filterMu.Unlock()
 			if filterCopy != nil {
 				return filterCopy(filterArgs)
 			}
 			return nil
 		}
-	var waitForTransferBlocked atomic.Value
-	waitForTransferBlocked.Store(false)
-	transferBlocked := make(chan struct{})
+
+	l.waitForTransferBlocked.Store(false)
+	l.transferBlocked = make(chan struct{})
 	cfg.TestingKnobs.LeaseTransferBlockedOnExtensionEvent = func(
 		_ roachpb.ReplicaDescriptor) {
-		if waitForTransferBlocked.Load().(bool) {
-			transferBlocked <- struct{}{}
-			waitForTransferBlocked.Store(false)
+		if l.waitForTransferBlocked.Load().(bool) {
+			l.transferBlocked <- struct{}{}
+			l.waitForTransferBlocked.Store(false)
 		}
 	}
-	mtc := &multiTestContext{}
-	mtc.storeConfig = &cfg
-	defer mtc.Stop()
-	mtc.Start(t, 2)
-	mtc.initGossipNetwork()
+
+	l.mtc = &multiTestContext{}
+	l.mtc.storeConfig = &cfg
+	l.mtc.Start(t, 2)
+	l.mtc.initGossipNetwork()
 
 	// First, do a write; we'll use it to determine when the dust has settled.
-	leftKey := roachpb.Key("a")
-	incArgs := incrementArgs(leftKey, 1)
-	if _, pErr := client.SendWrapped(context.Background(), mtc.distSenders[0], incArgs); pErr != nil {
+	l.leftKey = roachpb.Key("a")
+	incArgs := incrementArgs(l.leftKey, 1)
+	if _, pErr := client.SendWrapped(context.Background(), l.mtc.distSenders[0], incArgs); pErr != nil {
 		t.Fatal(pErr)
 	}
 
 	// Get the left range's ID.
-	rangeID := mtc.stores[0].LookupReplica(roachpb.RKey("a"), nil).RangeID
+	rangeID := l.mtc.stores[0].LookupReplica(keys.MustAddr(l.leftKey), nil).RangeID
 
 	// Replicate the left range onto node 1.
-	mtc.replicateRange(rangeID, 1)
+	l.mtc.replicateRange(rangeID, 1)
 
-	replicas := make([]*storage.Replica, len(mtc.stores))
-	replicaDescs := make([]roachpb.ReplicaDescriptor, len(mtc.stores))
-	for i, store := range mtc.stores {
+	l.replica0 = l.mtc.stores[0].LookupReplica(roachpb.RKey("a"), nil)
+	l.replica1 = l.mtc.stores[1].LookupReplica(roachpb.RKey("a"), nil)
+	{
 		var err error
-		replicas[i] = store.LookupReplica(roachpb.RKey("a"), nil)
-		replicaDescs[i], err = replicas[i].GetReplicaDescriptor()
-		if err != nil {
+		if l.replica0Desc, err = l.replica0.GetReplicaDescriptor(); err != nil {
+			t.Fatal(err)
+		}
+		if l.replica1Desc, err = l.replica1.GetReplicaDescriptor(); err != nil {
 			t.Fatal(err)
 		}
 	}
 
-	sendRead := func(storeIdx int) *roachpb.Error {
-		_, pErr := client.SendWrappedWith(
-			context.Background(),
-			mtc.senders[storeIdx],
-			roachpb.Header{RangeID: rangeID, Replica: replicaDescs[storeIdx]},
-			getArgs(leftKey),
-		)
-		return pErr
-	}
-
 	// Check that replica0 can serve reads OK.
-	if pErr := sendRead(0); pErr != nil {
+	if pErr := l.sendRead(0); pErr != nil {
 		t.Fatal(pErr)
 	}
+	return l
+}
 
-	// checkHasLease checks that a lease for the left range is owned by a
-	// replica. The check is executed in a retry loop because the lease may not
-	// have been applied yet.
-	checkHasLease := func(t *testing.T, storeIdx int) {
-		t.Helper()
-		testutils.SucceedsSoon(t, func() error {
-			return sendRead(storeIdx).GoError()
-		})
+func (l *leaseTransferTest) sendRead(storeIdx int) *roachpb.Error {
+	desc := l.mtc.stores[storeIdx].LookupReplica(keys.MustAddr(l.leftKey), nil)
+	replicaDesc, err := desc.GetReplicaDescriptor()
+	if err != nil {
+		return roachpb.NewError(err)
 	}
+	_, pErr := client.SendWrappedWith(
+		context.Background(),
+		l.mtc.senders[storeIdx],
+		roachpb.Header{RangeID: desc.RangeID, Replica: replicaDesc},
+		getArgs(l.leftKey),
+	)
+	if pErr != nil {
+		log.Warning(context.TODO(), pErr)
+	}
+	return pErr
+}
 
-	// setFilter is a helper function to enable/disable the blocking of
-	// RequestLeaseRequests on replica1. This function will notify that an
-	// extension is blocked on the passed in channel and will wait on the same
-	// channel to unblock the extension. Note that once an extension is blocked,
-	// the filter is cleared.
-	setFilter := func(setTo bool, extensionSem chan struct{}) {
-		filterMu.Lock()
-		defer filterMu.Unlock()
-		if !setTo {
-			filter = nil
-			return
-		}
-		filter = func(filterArgs storagebase.FilterArgs) *roachpb.Error {
-			if filterArgs.Sid != mtc.stores[1].Ident.StoreID {
-				return nil
-			}
-			llReq, ok := filterArgs.Req.(*roachpb.RequestLeaseRequest)
-			if !ok {
-				return nil
-			}
-			if llReq.Lease.Replica == replicaDescs[1] {
-				// Notify the main thread that the extension is in progress and wait for
-				// the signal to proceed.
-				filterMu.Lock()
-				filter = nil
-				filterMu.Unlock()
-				extensionSem <- struct{}{}
-				<-extensionSem
-			}
+// checkHasLease checks that a lease for the left range is owned by a
+// replica. The check is executed in a retry loop because the lease may not
+// have been applied yet.
+func (l *leaseTransferTest) checkHasLease(t *testing.T, storeIdx int) {
+	t.Helper()
+	testutils.SucceedsSoon(t, func() error {
+		return l.sendRead(storeIdx).GoError()
+	})
+}
+
+// setFilter is a helper function to enable/disable the blocking of
+// RequestLeaseRequests on replica1. This function will notify that an
+// extension is blocked on the passed in channel and will wait on the same
+// channel to unblock the extension. Note that once an extension is blocked,
+// the filter is cleared.
+func (l *leaseTransferTest) setFilter(setTo bool, extensionSem chan struct{}) {
+	l.filterMu.Lock()
+	defer l.filterMu.Unlock()
+	if !setTo {
+		l.filter = nil
+		return
+	}
+	l.filter = func(filterArgs storagebase.FilterArgs) *roachpb.Error {
+		if filterArgs.Sid != l.mtc.stores[1].Ident.StoreID {
 			return nil
 		}
-	}
-
-	forceLeaseExtension := func(storeIdx int, lease roachpb.Lease) error {
-		shouldRenewTS := lease.Expiration.Add(-1, 0)
-		mtc.manualClock.Set(shouldRenewTS.WallTime + 1)
-		err := sendRead(storeIdx).GoError()
-		if err != nil {
-			// We can sometimes receive an error from our renewal attempt because the
-			// lease transfer ends up causing the renewal to re-propose and second
-			// attempt fails because it's already been renewed. This used to work
-			// before we compared the proposer's lease with the actual lease because
-			// the renewed lease still encompassed the previous request.
-			if _, ok := err.(*roachpb.NotLeaseHolderError); ok {
-				err = nil
-			}
+		llReq, ok := filterArgs.Req.(*roachpb.RequestLeaseRequest)
+		if !ok {
+			return nil
 		}
-		return err
+		if llReq.Lease.Replica == l.replica1Desc {
+			// Notify the main thread that the extension is in progress and wait for
+			// the signal to proceed.
+			l.filterMu.Lock()
+			l.filter = nil
+			l.filterMu.Unlock()
+			extensionSem <- struct{}{}
+			<-extensionSem
+		}
+		return nil
 	}
+}
+
+func (l *leaseTransferTest) forceLeaseExtension(storeIdx int, lease roachpb.Lease) error {
+	shouldRenewTS := lease.Expiration.Add(-1, 0)
+	l.mtc.manualClock.Set(shouldRenewTS.WallTime + 1)
+	err := l.sendRead(storeIdx).GoError()
+	if err != nil {
+		// We can sometimes receive an error from our renewal attempt because the
+		// lease transfer ends up causing the renewal to re-propose and second
+		// attempt fails because it's already been renewed. This used to work
+		// before we compared the proposer's lease with the actual lease because
+		// the renewed lease still encompassed the previous request.
+		if _, ok := err.(*roachpb.NotLeaseHolderError); ok {
+			err = nil
+		}
+	}
+	return err
+}
+
+// ensureLeaderAndRaftState is a helper function that blocks until leader is
+// the raft leader and follower is up to date.
+func (l *leaseTransferTest) ensureLeaderAndRaftState(
+	t *testing.T, leader *storage.Replica, follower roachpb.ReplicaDescriptor,
+) {
+	t.Helper()
+	leaderDesc, err := leader.GetReplicaDescriptor()
+	if err != nil {
+		t.Fatal(err)
+	}
+	testutils.SucceedsSoon(t, func() error {
+		r := l.mtc.getRaftLeader(l.replica0.RangeID)
+		if r == nil {
+			return errors.Errorf("could not find raft leader replica for range %d", l.replica0.RangeID)
+		}
+		desc, err := r.GetReplicaDescriptor()
+		if err != nil {
+			return errors.Wrap(err, "could not get replica descriptor")
+		}
+		if desc != leaderDesc {
+			return errors.Errorf(
+				"expected replica with id %v to be raft leader, instead got id %v",
+				leaderDesc.ReplicaID,
+				desc.ReplicaID,
+			)
+		}
+		return nil
+	})
+
+	testutils.SucceedsSoon(t, func() error {
+		status := leader.RaftStatus()
+		progress, ok := status.Progress[uint64(follower.ReplicaID)]
+		if !ok {
+			return errors.Errorf(
+				"replica %v progress not found in progress map: %v",
+				follower.ReplicaID,
+				status.Progress,
+			)
+		}
+		if progress.Match < status.Commit {
+			return errors.Errorf("replica %v failed to catch up", follower.ReplicaID)
+		}
+		return nil
+	})
+}
+
+func TestRangeTransferLeaseExpirationBased(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
 	t.Run("Transfer", func(t *testing.T) {
-		origLease, _ := replicas[0].GetLease()
+		l := setupLeaseTransferTest(t)
+		defer l.mtc.Stop()
+		origLease, _ := l.replica0.GetLease()
 		{
 			// Transferring the lease to ourself should be a no-op.
-			if err := replicas[0].AdminTransferLease(context.Background(), replicaDescs[0].StoreID); err != nil {
+			if err := l.replica0.AdminTransferLease(context.Background(), l.replica0Desc.StoreID); err != nil {
 				t.Fatal(err)
 			}
-			newLease, _ := replicas[0].GetLease()
+			newLease, _ := l.replica0.GetLease()
 			if !origLease.Equivalent(newLease) {
 				t.Fatalf("original lease %v and new lease %v not equivalent", origLease, newLease)
 			}
@@ -639,30 +714,30 @@ func TestRangeTransferLease(t *testing.T) {
 		{
 			// An invalid target should result in an error.
 			const expected = "unable to find store .* in range"
-			if err := replicas[0].AdminTransferLease(context.Background(), 1000); !testutils.IsError(err, expected) {
+			if err := l.replica0.AdminTransferLease(context.Background(), 1000); !testutils.IsError(err, expected) {
 				t.Fatalf("expected %s, but found %v", expected, err)
 			}
 		}
 
-		if err := replicas[0].AdminTransferLease(context.Background(), replicaDescs[1].StoreID); err != nil {
+		if err := l.replica0.AdminTransferLease(context.Background(), l.replica1Desc.StoreID); err != nil {
 			t.Fatal(err)
 		}
 
-		// Check that replicas[0] doesn't serve reads any more.
-		pErr := sendRead(0)
+		// Check that replica0 doesn't serve reads any more.
+		pErr := l.sendRead(0)
 		nlhe, ok := pErr.GetDetail().(*roachpb.NotLeaseHolderError)
 		if !ok {
 			t.Fatalf("expected %T, got %s", &roachpb.NotLeaseHolderError{}, pErr)
 		}
-		if *(nlhe.LeaseHolder) != replicaDescs[1] {
+		if *(nlhe.LeaseHolder) != l.replica1Desc {
 			t.Fatalf("expected lease holder %+v, got %+v",
-				replicaDescs[1], nlhe.LeaseHolder)
+				l.replica1Desc, nlhe.LeaseHolder)
 		}
 
 		// Check that replica1 now has the lease.
-		checkHasLease(t, 1)
+		l.checkHasLease(t, 1)
 
-		replica1Lease, _ := replicas[1].GetLease()
+		replica1Lease, _ := l.replica1.GetLease()
 
 		// We'd like to verify the timestamp cache's low water mark, but this is
 		// impossible to determine precisely in all cases because it may have
@@ -670,7 +745,7 @@ func TestRangeTransferLease(t *testing.T) {
 		// low water mark, we make sure that the high water mark is equal to or
 		// greater than the new lease start time, which is less than the
 		// previous lease's expiration time.
-		if highWater := replicas[1].GetTSCacheHighWater(); highWater.Less(replica1Lease.Start) {
+		if highWater := l.replica1.GetTSCacheHighWater(); highWater.Less(replica1Lease.Start) {
 			t.Fatalf("expected timestamp cache high water %s, but found %s",
 				replica1Lease.Start, highWater)
 		}
@@ -680,30 +755,32 @@ func TestRangeTransferLease(t *testing.T) {
 	// that. Test that the transfer still happens (it'll wait until the extension
 	// is done).
 	t.Run("TransferWithExtension", func(t *testing.T) {
+		l := setupLeaseTransferTest(t)
+		defer l.mtc.Stop()
 		// Ensure that replica1 has the lease.
-		if err := replicas[0].AdminTransferLease(context.Background(), replicaDescs[1].StoreID); err != nil {
+		if err := l.replica0.AdminTransferLease(context.Background(), l.replica1Desc.StoreID); err != nil {
 			t.Fatal(err)
 		}
-		checkHasLease(t, 1)
+		l.checkHasLease(t, 1)
 
 		extensionSem := make(chan struct{})
-		setFilter(true, extensionSem)
+		l.setFilter(true, extensionSem)
 
 		// Initiate an extension.
 		renewalErrCh := make(chan error)
 		go func() {
-			lease, _ := replicas[1].GetLease()
-			renewalErrCh <- forceLeaseExtension(1, lease)
+			lease, _ := l.replica1.GetLease()
+			renewalErrCh <- l.forceLeaseExtension(1, lease)
 		}()
 
 		// Wait for extension to be blocked.
 		<-extensionSem
-		waitForTransferBlocked.Store(true)
+		l.waitForTransferBlocked.Store(true)
 		// Initiate a transfer.
 		transferErrCh := make(chan error)
 		go func() {
-			// Transfer back from replica1 to replicas[0].
-			err := replicas[1].AdminTransferLease(context.Background(), replicaDescs[0].StoreID)
+			// Transfer back from replica1 to replica0.
+			err := l.replica1.AdminTransferLease(context.Background(), l.replica0Desc.StoreID)
 			// Ignore not leaseholder errors which can arise due to re-proposals.
 			if _, ok := err.(*roachpb.NotLeaseHolderError); ok {
 				err = nil
@@ -711,11 +788,11 @@ func TestRangeTransferLease(t *testing.T) {
 			transferErrCh <- err
 		}()
 		// Wait for the transfer to be blocked by the extension.
-		<-transferBlocked
+		<-l.transferBlocked
 		// Now unblock the extension.
 		extensionSem <- struct{}{}
-		checkHasLease(t, 0)
-		setFilter(false, nil)
+		l.checkHasLease(t, 0)
+		l.setFilter(false, nil)
 
 		if err := <-renewalErrCh; err != nil {
 			t.Errorf("unexpected error from lease renewal: %s", err)
@@ -725,111 +802,71 @@ func TestRangeTransferLease(t *testing.T) {
 		}
 	})
 
-	// ensureLeaderAndRaftState is a helper function that blocks until leader is
-	// the raft leader and follower is up to date.
-	ensureLeaderAndRaftState := func(leader *storage.Replica, follower roachpb.ReplicaDescriptor) {
-		t.Helper()
-		leaderDesc, err := leader.GetReplicaDescriptor()
-		if err != nil {
-			t.Fatal(err)
-		}
-		testutils.SucceedsSoon(t, func() error {
-			r := mtc.getRaftLeader(rangeID)
-			if r == nil {
-				return errors.Errorf("could not find raft leader replica for range %d", rangeID)
-			}
-			desc, err := r.GetReplicaDescriptor()
-			if err != nil {
-				return errors.Wrap(err, "could not get replica descriptor")
-			}
-			if desc != leaderDesc {
-				return errors.Errorf(
-					"expected replica with id %v to be raft leader, instead got id %v",
-					leaderDesc.ReplicaID,
-					desc.ReplicaID,
-				)
-			}
-			return nil
-		})
-
-		testutils.SucceedsSoon(t, func() error {
-			status := leader.RaftStatus()
-			progress, ok := status.Progress[uint64(follower.ReplicaID)]
-			if !ok {
-				return errors.Errorf(
-					"replica %v progress not found in progress map: %v",
-					follower.ReplicaID,
-					status.Progress,
-				)
-			}
-			if progress.Match < status.Commit {
-				return errors.Errorf("replica %v failed to catch up", follower.ReplicaID)
-			}
-			return nil
-		})
-	}
-
 	// DrainTransfer verifies that a draining store attempts to transfer away
 	// range leases owned by its replicas.
 	t.Run("DrainTransfer", func(t *testing.T) {
+		l := setupLeaseTransferTest(t)
+		defer l.mtc.Stop()
 		// We have to ensure that replica0 is the raft leader and that replica1 has
 		// caught up to replica0 as draining code doesn't transfer leases to
 		// behind replicas.
-		ensureLeaderAndRaftState(replicas[0], replicaDescs[1])
-		mtc.stores[0].SetDraining(true)
+		l.ensureLeaderAndRaftState(t, l.replica0, l.replica1Desc)
+		l.mtc.stores[0].SetDraining(true)
 
 		// Check that replica0 doesn't serve reads any more.
-		pErr := sendRead(0)
+		pErr := l.sendRead(0)
 		nlhe, ok := pErr.GetDetail().(*roachpb.NotLeaseHolderError)
 		if !ok {
 			t.Fatalf("expected %T, got %s", &roachpb.NotLeaseHolderError{}, pErr)
 		}
-		if nlhe.LeaseHolder == nil || *nlhe.LeaseHolder != replicaDescs[1] {
+		if nlhe.LeaseHolder == nil || *nlhe.LeaseHolder != l.replica1Desc {
 			t.Fatalf("expected lease holder %+v, got %+v",
-				replicaDescs[1], nlhe.LeaseHolder)
+				l.replica1Desc, nlhe.LeaseHolder)
 		}
 
 		// Check that replica1 now has the lease.
-		checkHasLease(t, 1)
+		l.checkHasLease(t, 1)
 
-		mtc.stores[0].SetDraining(false)
+		l.mtc.stores[0].SetDraining(false)
 	})
 
 	// DrainTransferWithExtension verifies that a draining store waits for any
 	// in-progress lease requests to complete before transferring away the new
 	// lease.
 	t.Run("DrainTransferWithExtension", func(t *testing.T) {
+		l := setupLeaseTransferTest(t)
+		defer l.mtc.Stop()
 		// Ensure that replica1 has the lease.
-		if err := replicas[0].AdminTransferLease(context.Background(), replicaDescs[1].StoreID); err != nil {
+		if err := l.replica0.AdminTransferLease(context.Background(), l.replica1Desc.StoreID); err != nil {
 			t.Fatal(err)
 		}
-		checkHasLease(t, 1)
+		l.checkHasLease(t, 1)
 
 		extensionSem := make(chan struct{})
-		setFilter(true, extensionSem)
+		l.setFilter(true, extensionSem)
 
 		// Initiate an extension.
 		renewalErrCh := make(chan error)
 		go func() {
-			lease, _ := replicas[1].GetLease()
-			renewalErrCh <- forceLeaseExtension(1, lease)
+			lease, _ := l.replica1.GetLease()
+			renewalErrCh <- l.forceLeaseExtension(1, lease)
 		}()
 
 		// Wait for extension to be blocked.
 		<-extensionSem
 
 		// Make sure that replica 0 is up to date enough to receive the lease.
-		ensureLeaderAndRaftState(replicas[1], replicaDescs[0])
+		l.ensureLeaderAndRaftState(t, l.replica1, l.replica0Desc)
 
 		// Drain node 1 with an extension in progress.
 		go func() {
-			mtc.stores[1].SetDraining(true)
+			l.mtc.stores[1].SetDraining(true)
 		}()
 		// Now unblock the extension.
 		extensionSem <- struct{}{}
 
-		checkHasLease(t, 0)
-		setFilter(false, nil)
+		l.checkHasLease(t, 0)
+		l.setFilter(false, nil)
 
 		if err := <-renewalErrCh; err != nil {
 			t.Errorf("unexpected error from lease renewal: %s", err)

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -4015,6 +4015,7 @@ func (r *Replica) hasRaftReadyRLocked() bool {
 	return r.mu.internalRaftGroup.HasReady()
 }
 
+// TODO(tschottdorf): there's also maybeTransferRaftLeadership. Only one should exist.
 func (r *Replica) maybeTransferRaftLeader(
 	ctx context.Context, status *raft.Status, now hlc.Timestamp,
 ) {

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -419,6 +419,8 @@ func addSSTablePreApply(
 // away from this node to target, if this node is the current raft
 // leader. We don't attempt to transfer leadership if the transferee
 // is behind on applying the log.
+//
+// TODO(tschottdorf): there's also maybeTransferRaftLeader. Only one should exist.
 func (r *Replica) maybeTransferRaftLeadership(ctx context.Context, target roachpb.ReplicaID) {
 	err := r.withRaftGroup(func(raftGroup *raft.RawNode) (bool, error) {
 		// Only the raft leader can attempt a leadership transfer.

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -959,6 +959,8 @@ func (s *Store) AnnotateCtx(ctx context.Context) context.Context {
 	return s.cfg.AmbientCtx.AnnotateCtx(ctx)
 }
 
+// The maximum amount of time waited for leadership shedding before commencing
+// to drain a store.
 const raftLeadershipTransferWait = 5 * time.Second
 
 // SetDraining (when called with 'true') causes incoming lease transfers to be
@@ -980,83 +982,132 @@ func (s *Store) SetDraining(drain bool) {
 	var wg sync.WaitGroup
 
 	ctx := log.WithLogTag(context.Background(), "drain", nil)
-	// Limit the number of concurrent lease transfers.
-	sem := make(chan struct{}, 100)
-	sysCfg, sysCfgSet := s.cfg.Gossip.GetSystemConfig()
-	newStoreReplicaVisitor(s).Visit(func(r *Replica) bool {
-		wg.Add(1)
-		if err := s.stopper.RunLimitedAsyncTask(
-			r.AnnotateCtx(ctx), "storage.Store: draining replica", sem, true, /* wait */
-			func(ctx context.Context) {
-				defer wg.Done()
+	transferAllAway := func() int {
+		// Limit the number of concurrent lease transfers.
+		sem := make(chan struct{}, 100)
+		sysCfg, sysCfgSet := s.cfg.Gossip.GetSystemConfig()
+		// Incremented for every lease or Raft leadership transfer attempted. We try
+		// to send both the lease and the Raft leaders away, but this may not
+		// reliably work. Instead, we run the surrounding retry loop until there are
+		// no leaders/leases left (ignoring single-replica or uninitialized Raft
+		// groups).
+		var numTransfersAttempted int32
+		newStoreReplicaVisitor(s).Visit(func(r *Replica) bool {
+			wg.Add(1)
+			if err := s.stopper.RunLimitedAsyncTask(
+				r.AnnotateCtx(ctx), "storage.Store: draining replica", sem, true, /* wait */
+				func(ctx context.Context) {
+					defer wg.Done()
 
-				r.mu.Lock()
-				r.mu.draining = true
-				r.mu.Unlock()
-
-				var drainingLease roachpb.Lease
-				for {
-					var llHandle *leaseRequestHandle
 					r.mu.Lock()
-					lease, nextLease := r.getLeaseRLocked()
-					if nextLease != nil && nextLease.OwnedBy(s.StoreID()) {
-						llHandle = r.mu.pendingLeaseRequest.JoinRequest()
-					}
+					r.mu.draining = true
+					status := r.raftStatusRLocked()
+					// needsRaftTransfer is true when we can reasonably hope to transfer
+					// this replica's lease and/or Raft leadership away.
+					needsRaftTransfer := status != nil &&
+						len(status.Progress) > 1 &&
+						!(status.RaftState == raft.StateFollower && status.Lead != 0)
 					r.mu.Unlock()
 
-					if llHandle != nil {
-						<-llHandle.C()
-						continue
-					}
-					drainingLease = lease
-					break
-				}
-
-				if drainingLease.OwnedBy(s.StoreID()) && r.IsLeaseValid(drainingLease, s.Clock().Now()) {
-					desc := r.Desc()
-					zone := config.DefaultZoneConfig()
-					if sysCfgSet {
-						var err error
-						zone, err = sysCfg.GetZoneConfigForKey(desc.StartKey)
-						if log.V(1) && err != nil {
-							log.Errorf(ctx, "could not get zone config for key %s when draining: %s", desc.StartKey, err)
+					var drainingLease roachpb.Lease
+					for {
+						var llHandle *leaseRequestHandle
+						r.mu.Lock()
+						lease, nextLease := r.getLeaseRLocked()
+						if nextLease != nil && nextLease.OwnedBy(s.StoreID()) {
+							llHandle = r.mu.pendingLeaseRequest.JoinRequest()
 						}
+						r.mu.Unlock()
+
+						if llHandle != nil {
+							<-llHandle.C()
+							continue
+						}
+						drainingLease = lease
+						break
 					}
-					transferred, err := s.replicateQueue.transferLease(
-						ctx,
-						r,
-						desc,
-						zone,
-						transferLeaseOptions{},
-					)
-					if log.V(1) {
-						if transferred {
-							log.Infof(ctx, "transferred lease %s for replica %s", drainingLease, desc)
-						} else {
+
+					ownsValidLease := drainingLease.OwnedBy(s.StoreID()) && r.IsLeaseValid(drainingLease, s.Clock().Now())
+
+					if ownsValidLease || needsRaftTransfer {
+						atomic.AddInt32(&numTransfersAttempted, 1)
+					}
+
+					if ownsValidLease {
+						desc := r.Desc()
+						zone := config.DefaultZoneConfig()
+						if sysCfgSet {
+							var err error
+							zone, err = sysCfg.GetZoneConfigForKey(desc.StartKey)
+							if log.V(1) && err != nil {
+								log.Errorf(ctx, "could not get zone config for key %s when draining: %s", desc.StartKey, err)
+							}
+						}
+						leaseTransferred, err := s.replicateQueue.transferLease(
+							ctx,
+							r,
+							desc,
+							zone,
+							transferLeaseOptions{},
+						)
+						if log.V(1) && !leaseTransferred {
 							// Note that a nil error means that there were no suitable
 							// candidates.
 							log.Errorf(
 								ctx,
-								"did not transfer lease %s for replica %s when draining: %s",
+								"did not transfer lease %s for replica %s when draining: %v",
 								drainingLease,
 								desc,
 								err,
 							)
 						}
+						if err == nil && leaseTransferred {
+							// If we just transferred the lease away, Raft leadership will
+							// usually transfer with it. Invoking a separate Raft leadership
+							// transfer would only obstruct this.
+							needsRaftTransfer = false
+						}
 					}
+
+					if needsRaftTransfer {
+						r.raftMu.Lock()
+						r.maybeTransferRaftLeadership(ctx, drainingLease.Replica.ReplicaID)
+						r.raftMu.Unlock()
+					}
+				}); err != nil {
+				if log.V(1) {
+					log.Errorf(ctx, "error running draining task: %s", err)
 				}
-			}); err != nil {
-			if log.V(1) {
-				log.Errorf(ctx, "error running draining task: %s", err)
+				wg.Done()
+				return false
 			}
-			wg.Done()
-			return false
+			return true
+		})
+		wg.Wait()
+		return int(numTransfersAttempted)
+	}
+
+	transferAllAway()
+
+	var cancel func()
+	ctx, cancel = context.WithTimeout(ctx, raftLeadershipTransferWait)
+	defer cancel()
+
+	opts := retry.Options{
+		InitialBackoff: 10 * time.Millisecond,
+		MaxBackoff:     time.Second,
+		Multiplier:     2,
+	}
+	// Avoid retry.ForDuration because of https://github.com/cockroachdb/cockroach/issues/25091.
+	if err := retry.WithMaxAttempts(ctx, opts, 10000, func() error {
+		if numRemaining := transferAllAway(); numRemaining > 0 {
+			return errors.Errorf("waiting for %d replicas to transfer their lease away", numRemaining)
 		}
-		return true
-	})
-	wg.Wait()
-	if drain {
-		time.Sleep(raftLeadershipTransferWait)
+		return nil
+	}); err != nil {
+		// You expect this message when shutting down a server in an unhealthy
+		// cluster. If we see it on healthy ones, there's likely something to fix.
+		log.Warningf(ctx, "unable to drain cleanly within %s, service might briefly deteriorate: %s", raftLeadershipTransferWait, err)
 	}
 }
 

